### PR TITLE
Flexible Federated Learning government architecture

### PR DESCRIPTION
**Pull request by gstipcich**
The main contribution is the new architecture that allows more flexible definition of the FL Government. 
In this setting, the user can easily define the *federated government round* by orchestrating the operations between the nodes of the federation. 
The nodes of the federation can be any: typically, the server and the client nodes are considered. However, in general, a node can have any private data (labels), its own model, and access policies. 
This setting allows to naturally accommodate Vertical and Transfer FL algorithms. 
One architecture for Vertical FL using neural networks is included in the source code.  

Other contributions are refactoring to allow an easier and more intuitive usage of the code, which in turn reduces the effort to define customized functions.

Moreover, Pylint has been passed to the whole source code and tests, and the documentation have been updated to comply to Google's style guide.

### CODE REVIEW

File: shfl/model/vertical_deep_learning_model.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+                self._epoch_counter += 1
+
+    def get_meta_params(self):
+        """Returns the computed embeddings and associated indices."""
+
+        return self._embeddings.detach().cpu().numpy(), \
+            self._embeddings_indices
+
+
+class VerticalNeuralNetServer(DeepLearningModelPyTorch):
```
Comment (gegose) [inline shfl/model/vertical_deep_learning_model.py:93]: It migth be redundant but in this case(and in the previous one) i think VerticalNeuralNetServerModel could be a better name because when i read ....Server i always think in a machine, don't you?

File: shfl/private/utils.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    # Arguments:
+        data: Input data.
+        mean: Mean used for the normalization.
+        std: Standard deviation used for the normalization.
+    """
+    data.data = (data.data - mean) / std
+
+
+def shuffle_node_query(data):
```
[the remark this answers is not in the document]
Comment (gstipcich) [reply shfl/private/utils.py:26]: I definitely agree: these are simple functions that can be actually defined with a lambda. 
We could: 
1) Remove the functions from the source code
2) Edit the few notebooks that were using them.

File: shfl/private/federated_attack.py [hunk 3]
```diff
@@ -1,76 +1,48 @@
...
         return self._adversaries
 
-    def apply_attack(self, federated_data):
-        """
-        Method that implements federated attack of data poisoning shuffling training labels of some nodes.
+    def __call__(self, nodes_federation):
+        """Shuffles the target labels in an adversary node.
+
+        If tagged as adversarial, a node's target label is shuffled
+        by a data transformation (see [ShuffleNode](./#shufflenode-class)).
```
[the remark this answers is not in the document]
Comment (gstipcich) [reply shfl/private/federated_attack.py:33]: Since this was a child class of `FederatedTransformation`, which we decided to turn into a function, I turned the PoisoningAttack into a callable as well (and removed the abstract class).

File: shfl/model/vertical_deep_learning_model.py (hunk 1 above)
Comment (gstipcich) [reply shfl/model/vertical_deep_learning_model.py:93]: I definitely agree: this should be renamed. Also because it contains a Neural Model.

File: shfl/private/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gstipcich) [reply shfl/private/utils.py:26]: A small remark: we can use the delta function as a query in the cases where there is no assignment to the data (like in `random.shuffle`, where the shuffle is "in place"). 

However, for other queries (such as normalization or reshape), we need to actually write a simple function where there is an assignment to the data. For example:

```{python}
def normalize_data(data, mean, std):
    data.data = (data.data - mean) / std
```
This remark does not change my opinion: we can safely remove those small functions from the source code, and let the user define the ones she/he needs in the specific situation. 

What do you think?

[the remark this answers is not in the document]
Comment (gegose) [reply shfl/private/utils.py:26]: Exactly, lambdas are not always the best option, but the point with meaningless functions is the same.

**Review by gegose — APPROVED** [review]
